File: cpp/euler145.cpp
/* Problem 145
 *
 * Some positive integers n have the property that the sum [ n + reverse(n) ]
 * consists entirely of odd (decimal) digits. For instance,
 * 36 + 63 = 99 and 409 + 904 = 1313. We will call such numbers reversible;
 * so 36, 63, 409, and 904 are reversible. Leading zeroes are not allowed in
 * either n or reverse(n).
 *
 * There are 120 reversible numbers below one-thousand.
 *
 * How many reversible numbers are there below one-billion (10^(9))?
 *
 */

#include <iostream>

template <typename T>
inline T reverse(T n)
{
	T r = 0;
	while (n > 0)
	{
		r = r + n % 10;
		n /= 10;
		r *= 10;
	}
	return r / 10;
}

template <typename T>
inline T is_all_odd(T n)
{
	for (; n > 0; n /= 10)
		if (!(n & 1)) return 0;
	return 1;
}

template <typename T>
inline T is_reversible(T n)
{
	if (n % 10 == 0) return 0;
	T r = reverse(n);
	if (r % 10 == 0) return 0;
	if ((r + n) & 1 == 0) return 0;
	else return is_all_odd(n + r);
}

int main(int argc, char *argv[])
{
	unsigned long long num_reversible = 0;
	for (unsigned long long i = 0; i < 1000000000; ++i)
		num_reversible += is_reversible(i);

	std::cout << "Answer is " << num_reversible << std::endl;

	return 0;
}

File: python/euler145.py
#!/usr/bin/env python

"""
Some positive integers n have the property that the sum [ n + reverse(n) ] consists entirely of odd (decimal) digits. For instance, 36 + 63 = 99 and 409 + 904 = 1313. We will call such numbers reversible; so 36, 63, 409, and 904 are reversible. Leading zeroes are not allowed in either n or reverse(n).

There are 120 reversible numbers below one-thousand.

How many reversible numbers are there below one-billion (10^(9))?
"""

from utils import reverse

def is_all_odd(n):
	while n > 0:
		if n & 1 == 0:
			return False
		n /= 10
	return True

def is_reversible(n):
	if n % 10 == 0:
		return False
	r = reverse(n)
	if r % 10 == 0:
		return False
	else:
		if r + n & 1 == 0:
			return False
		else:
			return is_all_odd(n + r)

i, num_reversible = 0, 0
while i < 1000000:
	num_reversible += is_reversible(i)
	i += 1

print "Answer is", num_reversible

"""
num_reversible = sum([is_reversible(i) for i in xrange(10**9)])
print "Answer is", num_reversible

givez
Macintosh:projecteuler jonasg$ time ./euler145.py 
Traceback (most recent call last):
  File "./euler145.py", line 32, in <module>
    num_reversible = sum([is_reversible(i) for i in xrange(10**9)])
MemoryError

real	78m0.968s
user	73m39.828s
sys	0m15.685s
"""
